backend/mem/location.go

```diff
@@ -0,0 +1,202 @@
+func (l *Location) String() string {
+
+	return l.URI()
+}
+
+/*
+List finds all of the files living at the current location and returns them in a slice of strings.
+If there are no files at location, then an empty slice will be returned
+*/
+
+func (l *Location) List() ([]string, error) {
+
+	str := l.Path()                         //full path of this location
+	mapRef := &l.fileSystem.fsMap           //setting mapRef to this value for code readability
+	if _, ok := (*mapRef)[l.Volume()]; ok { //are there paths on this volume?
+		list := (*mapRef)[l.Volume()].fileNamesHere(str) //getting a list of the file names on this location
+
+		return list, nil // "fileNamesHere" returns an empty list if no files were found
+	}
+	return make([]string, 0), nil //if the volume has nothing on it, return an empty list as well
+}
+
+/*
+ListByPrefix tags a prefix onto the current path and in a slice,
+ returns all file base names whose full paths contain that substring
+Returns empty slice if nothing found
+*/
+
+func (l *Location) ListByPrefix(prefix string) ([]string, error) {
+
+	list := make([]string, 0)
+	str := path.Join(l.Path(), prefix)
+	mapRef := &l.fileSystem.fsMap
+	if _, ok := (*mapRef)[l.volume]; ok {
+		paths := (*mapRef)[l.volume].getKeys()
+		for i := range paths {
+			if strings.Contains(paths[i], str) {
+				if path.Ext(paths[i]) != "" && strings.Contains(str, utils.EnsureTrailingSlash(path.Dir(paths[i]))) {
+					list = append(list, path.Base(paths[i]))
+					sort.Strings(list)
+				}
+			}
+		}
+	}
+
+	return list, nil
+}
+
+/*
+ListByRegex takes a regular expression and returns
+a slice of strings containing the base names of files
+found that matched the regular expression.  Returns an
+empty slice upon nothing found
+*/
+
+func (l *Location) ListByRegex(regex *regexp.Regexp) ([]string, error) {
+
+	list := make([]string, 0)
+	str := l.Path()
+	mapRef := &l.fileSystem.fsMap
+	if _, ok := (*mapRef)[l.Volume()]; ok {
+		namesHere := (*mapRef)[l.Volume()].fileNamesHere(str)
+		for i := range namesHere {
+			if regex.MatchString(namesHere[i]) {
+				list = append(list, namesHere[i])
+			}
+		}
+	}
+	return list, nil
+}
+
+//Volume returns the volume of the current FileSystem.
+func (l *Location) Volume() string {
+	return l.volume
+}
+
+//Path returns the full, absolute path of the location with leading and trailing slashes
+func (l *Location) Path() string {
+
+	//just to be sure that we return a trailing and leading slash
+	str := utils.EnsureTrailingSlash(path.Clean(l.name))
+	return utils.EnsureLeadingSlash(str)
+
+}
+
+//Exists always returns true on locations
+func (l *Location) Exists() (bool, error) {
+
+	l.exists = true
+	return true, nil
+}
+
+/*
+NewLocation creates a new location at the
+given relative path, which is tagged onto the current locations absolute path
+*/
+func (l *Location) NewLocation(relLocPath string) (vfs.Location, error) {
+	err := utils.ValidateRelativeLocationPath(relLocPath)
```

Comment:
This function should check the file system to see if a location already exist. Return early with that location's reference if it exists.